cmd/vault.go

```
@@ -397,3 +402,326 @@ func readJSON(name string) (secret interface{}, err error) {
+	}
+
+	flags := cmd.Flags()
+	flags.IntVar(&options.offset, "offset", 0, "where to start fetching the items")
+	flags.IntVar(&options.limit, "limit", 100, "number of items to return")
+	flags.StringVar(&options.ownerID, "user-id", "", "User ID of the user who owns the secret")
+	cmd.MarkFlagRequired("user-id")
+
+	cmd.AddCommand(userSecretShowCmd())
+	cmd.AddCommand(userSecretCreateCmd())
+	cmd.AddCommand(userSecretUpdateCmd())
+	cmd.AddCommand(userSecretDeleteCmd())
+	cmd.AddCommand(userSecretMetadataShowCmd())
+	return cmd
+}
+
+func userSecretList(options vaultOptions) error {
+	api := vault.New(curl())
+	secretID := vault.SecretID{
+		OwnerID: options.ownerID,
+		Name:    "",
+	}
+	secrets, err := api.UserSecrets(secretID, options.offset, options.limit)
+	if err != nil {
+		return err
+	}
+
+	return stdout(secrets)
+}
+
+//
+
+// show user secret details
+func userSecretShowCmd() *cobra.Command {
+	options := vaultOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "show",
+		Short: "Get a user secret",
+		Long:  `Get a user secret. Secret Name's are separated by commas when using multiple values, see example`,
+		Example: `
+	privx-cli user-secrets show [access flags] --name <SECRET-NAME>,<SECRET-NAME>
+		`,
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return userSecretShow(options)
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&options.secretName, "name", "", "secret name")
+	flags.StringVar(&options.ownerID, "owner-id", "", "secret's owner ID")
+	cmd.MarkFlagRequired("owner-id")
+	cmd.MarkFlagRequired("name")
+
+	return cmd
+}
+
+func userSecretShow(options vaultOptions) error {
+	api := vault.New(curl())
+	secrets := []vault.Secret{}
+	secretID := &vault.SecretID{
+		OwnerID: options.ownerID,
+		Name:    "",
+	}
+	for _, name := range strings.Split(options.secretName, ",") {
+		secretID.Name = name
+		secret, err := api.UserSecret(*secretID)
+		if err != nil {
+			return err
+		}
+		secrets = append(secrets, *secret)
```

probably over-optimisation when sharing the underlying Secret struct. It is OK to create new struct instance for each API call and to delegate the optimisation to the Golang GC. This makes it less error-prone to put things in parallel should it is needed in the future. Move the secretID into the for loop.

There are alternative error handlings that can be added:
- Filter out empty strings after splitting. 
- Check requested secret names length after filtering to keep the for loop sane.

---

Here another sensible optimisation (if must be done) is to pre-allocate the secrets slice.
```
// Instead of this
secrets := []vault.Secret{}
// Do this
secrets := make(vault.Secret, len(secretNames))
// ... where the secretNames is the cleaned input
```
However, since go-client has one-shot invocation it generally does not need to optimise memory usage except for very heavy resource. Thus, this suggested optimisation and the struct sharing are negligible and should be avoided to keep things simple.

---

Instead of going for optimisation, try improving UX aspect of the function. What if user provides 10 names and 1 one if them result in 404? With the current implementation an error is returned. If many are missing the user can only find out the error one by one.

This prevents the user from using the go-client effectively. How about adding --ignore-error to suppress errors and always return available data? The caller must check the resultant data as needed. The function will fetch as much data as it can in this mode.